### ISSUE TITLE
save_model fails to save the model after a certain number of rounds irrespective of the number of observations

### ISSUE THREAD

**Issue by supersaw-it** [catboost/catboost#2564]
Problem:
The `save_model` method fails to save the model after a certain number of rounds irrespective of the number of observations;
although the file itself is created, its size is 0 KB if `n_rounds` is set to a higher integer than _10910_. The `train_pool` dimensions are:
```
dim(training)
[1] 2647608     103
```

The successfully and empty saved models:

![catboo_n_rounds](https://github.com/catboost/catboost/assets/54410381/ebbcc536-6710-4118-a813-f441d966ba79)

However I would like to point out that I tried it with a lower and bigger dataset (+- 50% of the number of rows), it does not seem to influence the aforementioned outcome. I suspect it has something to do with model complexity, so as the model gets progressively bigger in size, it somehow fails to be saved after a certain threshold is reached.

Code:
```
import gc
import time
import pandas as pd
import numpy as np
from catboost import CatBoostClassifier, Pool
import os
import re

def change_to_latest_processed_dir():
    # Start from the root directory where the script is located
    root_directory = os.getcwd()

    # Define the path to the 'data' directory
    data_directory = os.path.join(root_directory, "data")

    # Check if the 'data' directory exists
    if not os.path.exists(data_directory):
        raise Exception("The 'data' directory does not exist in the root directory")

    # List all directories in the 'data' directory
    all_folders = [os.path.join(data_directory, d) for d in os.listdir(data_directory) if os.path.isdir(os.path.join(data_directory, d))]

    # Filter directories starting with "processed_2"
    processed_folders = [d for d in all_folders if re.match("^processed_2", os.path.basename(d))]

    # If there are no processed folders, stop the script
    if not processed_folders:
        raise Exception("No 'processed_2*' folders found in the 'data' directory")

    # Get the latest modified folder
    latest_folder = max(processed_folders, key=os.path.getmtime)

    # Change to the latest modified folder
    os.chdir(latest_folder)
    print(f"Changed directory to {latest_folder}")

# Call the function to change the directory
change_to_latest_processed_dir()

# Sample the middle bucket to lower the number of observations
trial = 1  # keep track of your work
var_to_predict = "ratio_return_group"
num_rounds = 10910  # 11500 fails
manual_cw = True
auto_cw = "None"  # SqrtBalanced / Balanced / None
metric = "MultiClass"  # MultiClass / MultiClassOneVsAll

# Load the pools
train_pool = Pool("cat_training.pool", column_description="cat_column_description.cd")

# Set seed for reproducibility
np.random.seed(777)

# Train CatBoost model
if manual_cw:
    class_weights = [1, 1, 1, 2, 2]  # Adjust as needed
    model = CatBoostClassifier(iterations=num_rounds, 
                            learning_rate=0.01, 
                            depth=12,
                            l2_leaf_reg=1.017,
                            subsample = 0.81,  # Sample rate for bagging (row subsampling)
                            bootstrap_type = 'Bernoulli',
                            loss_function=metric, 
                            eval_metric=metric, 
                            task_type='GPU',  # or 'CPU' if GPU support is not available
                            class_weights=class_weights)

else:
    model = CatBoostClassifier(iterations=num_rounds, 
                        learning_rate=0.01, 
                        depth=12,
                        l2_leaf_reg=1.017,
                        subsample = 0.81,  # Sample rate for bagging (row subsampling)
                        bootstrap_type = 'Bernoulli',
                        loss_function=metric, 
                        eval_metric=metric, 
                        task_type='GPU',  # or 'CPU' if GPU support is not available
                        auto_class_weights=auto_cw)

model.fit(train_pool)
time.sleep(3)
gc.collect()

class_weights_str = ''.join(map(str, class_weights)) if manual_cw else '0' # Convert class_weights to a string of concatenated numbers
# Construct the filename
model_filename = f"depth_trades_catboo_fact_trial_{trial}_nrounds_{num_rounds}_{var_to_predict}_manual_cw_{manual_cw}_{class_weights_str}_auto_cw_{auto_cw}_{metric}.cbm"

# Save the model
model.save_model(model_filename)
```
Setting either `auto_cw=True` or `class_weights` does not influence the outcome.
The result is consistent across 2 different machines with different GPUs and amount of RAM; also, both machines result in the same outcome regardless of the program running in Windows or WSL Debian.

PS The comparable R code results in the same outcome too.

Catboost version: 1.2.2

Operating System:
- Windows 11 Pro
- Debian GNU/Linux 12 (bookworm)

CPU: Intel(R) Core(TM) i9-14900K   3.20 GHz

GPU:
- NVIDIA GeForce RTX 3080  `# thought the problem could be my GPU / 16GB RAM, hence the upgrade to 64GB RAM & RTX 4090...`
- NVIDIA GeForce RTX 4090  `# 64GB RAM`

**Comment by andrey-khropov** [comment]
Indeed it seems that you're hitting a size limit of 2^31 bytes and the failure occurs after that.
It should be at least an exception, not a silent saving of the incorrect data. Although it might be possible to remove this limit, I'll look into it.

But I think it is likely not very practical to have such a big model. I suspect that you have at least some categorical features and some of them have many distinct values, right? You can find some tips on how to reduce the model size [here](https://catboost.ai/en/docs/concepts/faq#reduce-the-size-of-the-final-model), another possibility is to process categorical features as one-hot instead of computing CTRs, this is regulated by `one_hot_max_size` parameter, see [here](https://catboost.ai/en/docs/features/categorical-features).

**Comment by andrey-khropov** [comment]
> @andrey-khropov thanks for your reply! My dataset contains 1 categorical feature with 24 distinct values;

Hm. In this case I'd suspect that the model should not grow to such a big size.

> depth=12

And why do you use such a big depth? Is it the result of hyperparameter tuning? Such depth for oblivious trees is typically not a very good idea. Have you tried non-symmetric trees ([`grow_policy`](https://catboost.ai/en/docs/references/training-parameters/common#grow_policy) set to `Depthwise` or `Lossguide`)  ?

**Comment by supersaw-it** [comment]
@andrey-khropov could you clarify on the 'oblivious' trees, what does that mean?

Indeed it was a result of hyper-parameter tuning, however I have to confess I only had maybe 10 or 20 iterations, so it's probably not the best set of hyperparameters.

I haven't tried non-symmetric trees, will try that out, thanks!

**Comment by andrey-khropov** [comment]
> @andrey-khropov could you clarify on the 'oblivious' trees, what does that mean?

It means:
- All trees are [perfect binary trees](https://en.wikipedia.org/wiki/Binary_tree), it means that each such tree has $2^{depth}$ leaves, so this number grows exponentially with the increase of depth.
- All the decision predicates on each level of the tree are the same. 

It is the default mode of operation of CatBoost.

You can read some details, for example, [here](https://catboost.ai/news/catboost-enables-fast-gradient-boosting-on-decision-trees-using-gpus)

**Comment by supersaw-it** [comment]
@andrey-khropov thanks, very interesting, especially regarding the decision predicates!

I’ll experiment tomorrow; have a nice weekend.

**Comment by supersaw-it** [comment]
@andrey-khropov FYI I tried out `depth=11` as well as using `one_hot_max_size=2` (my only categorical variable with 24 distinct values **will not** be one-hot-encoded) & `one_hot_max_size=255` (my only categorical variable with 24 distinct values **will** be one-hot-encoded); moreover, I tried different tree growing policies:

![image](https://github.com/catboost/catboost/assets/54410381/215d1d6c-5133-47ea-b523-95182b7236eb)

At least when it comes to my dataset, only the depth significantly affects the size, which is expected.

I looked at the source code of Catboost and unfortunately could not find any mention of imposed file size limit when saving models. If such code exists, I would wholeheartedly appreciate if you or other contributors could point me to it;

Cheers!
